keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,289 @@
+[storageVersionHash][] field in the discovery document. The storage versions for
+built-in resources have been static. With this KEP, the kube-apiserver needs to
+dynamics update the storage version hash in the discovery document when the it
+gets the latest selected storage version from etcd.
+
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
+
+### CRDs
+
+Today, a [apiextension-apiserver][] uses informer to watch for the latest
+change to the CRD spec, and [builds][] new storage encoder with the latest storage
+version.
+
+The CRDs do not need the `StorageVersion` API to vote for a selected storage
+version. The storage version in the CRD spec is the "selected storage version".
+
+The current implementation of the apiextension-apiserver already provides the
+"loose guarantee".
+
+If we choose to pursue the "strict guarantee", most code changes will be done in
+the [etcd3 storage layer], which is shared by the kube-apiserver and
+apiextension-apiservers, so the CR storage will get the same enhancement. 
+
+[apiextension-apiserver]:https://github.com/kubernetes/kubernetes/blob/c91761da0d854aa842ea55ba4d7c6cbc8d675892/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L204
+[builds]:https://github.com/kubernetes/kubernetes/blob/c91761da0d854aa842ea55ba4d7c6cbc8d675892/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L678
+[etcd3 storage layer]:https://github.com/kubernetes/kubernetes/blob/74c0cc27902548243d8a863c9e2c6345a8e6b548/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go
+
+### Aggregated API servers
+
+Most code changes will be done in the generic apiserver library, so aggregated
+API servers using the library will get the same behavior.
+
+## Backwards Compatibility
+
+Clients do not see any behavioral change, so there is no backwards compatibility
+concern.
+
+## Graduation Plan
+
+* Alpha: in 1.16, the newly added API, including the `DiscoveryDocHashes` and
+  the `Consistent` field, will be feature gated by the
+  `EnableCoordinatedDiscoveryDocument` flag.
+
+* Beta & GA: if we don't find problems, we will graduate the API quarterly.
+
+## FAQ
+
+1. Q: how to prevent the following race between a client and the HA master:
+   client gets the discovery document, see storage version v1; HA master rolling
+   upgrades to use storage version v2; client sends a create request, expecting
+   the object to be encoded as v1 in etcd.
+
+   A: Unless API servers support transactions, the race is inevitable, and
+   exists in all parts of Kubernetes API. Client-side coordination is necessary
+   to overcome the race. Let's take the storage migrator as an example. For the
+   storage migrator to work properly, the required client-side coordination is
+   that at most one master version change could happen in any 5 minute window.
```

Comment:
i don't think this is *client-side coordination*, this is an assumption / guarantee made on the server side, i.e., cluster admins do *not* do version upgrade or downgrade more than once every 5 minutes.